File: test/test_system.js
const Database = require('../config/database');
const GoogleSpeechService = require('../config/google_speech');
const fs = require('fs');
const path = require('path');

class SystemTester {
    constructor() {
        this.database = new Database();
        this.speechService = new GoogleSpeechService();
    }

    async testDatabaseConnection() {
        console.log('Testing database connection...');
        try {
            if (this.database.connection) {
                console.log('✓ Database connection successful');
                
                // Test branch search
                const result = await this.database.getBranchExtension('กรุงเทพ');
                if (result) {
                    console.log(`✓ Branch search successful: ${JSON.stringify(result)}`);
                } else {
                    console.log('✗ Branch search failed');
                }
            } else {
                console.log('✗ Database connection failed');
            }
        } catch (error) {
            console.log(`✗ Database error: ${error.message}`);
        }
    }

    async testGoogleSpeech() {
        console.log('\nTesting Google Speech API...');
        try {
            // Test Text-to-Speech
            console.log('Testing Text-to-Speech...');
            const audioData = await this.speechService.textToSpeech('สวัสดีครับ ยินดีต้อนรับสู่ระบบ Call Center');
            if (audioData) {
                console.log('✓ Text-to-Speech successful');
                
                // Save audio file for testing
                const testFile = path.join(__dirname, 'test_audio.wav');
                await this.speechService.createAudioFile(audioData, testFile);
                console.log(`✓ Audio file saved: ${testFile}`);
            } else {
                console.log('✗ Text-to-Speech failed');
            }
            
            // Test Speech-to-Text (requires audio file)
            console.log('\nTesting Speech-to-Text...');
            console.log('Note: Speech-to-Text requires audio file input');
            
        } catch (error) {
            console.log(`✗ Google Speech error: ${error.message}`);
        }
    }

    async testAEPConfiguration() {
        console.log('\nTesting AEP configuration...');
        try {
            const configFiles = [
                '../asterisk/aeap.conf',
                '../asterisk/extensions.conf',
                '../asterisk/sip.conf',
                '../asterisk/manager.conf',
                '../asterisk/modules.conf'
            ];
            
            for (const configFile of configFiles) {
                const fullPath = path.join(__dirname, configFile);
                if (fs.existsSync(fullPath)) {
                    console.log(`✓ ${configFile} exists`);
                } else {
                    console.log(`✗ ${configFile} missing`);
                }
            }
        } catch (error) {
            console.log(`✗ AEP config error: ${error.message}`);
        }
    }

    async testNodeJSAGI() {
        console.log('\nTesting Node.js AGI script...');
        try {
            const agiPath = path.join(__dirname, '../agi_scripts/call_routing.js');
            if (fs.existsSync(agiPath)) {
                console.log('✓ AGI script exists');
                
                // Check if executable
                const stats = fs.statSync(agiPath);
                if (stats.mode & parseInt('111', 8)) {
                    console.log('✓ AGI script is executable');
                } else {
                    console.log('✗ AGI script is not executable');
                }
            } else {
                console.log('✗ AGI script not found');
            }
        } catch (error) {
            console.log(`✗ AGI script error: ${error.message}`);
        }
    }

    async testPackageDependencies() {
        console.log('\nTesting package dependencies...');
        try {
            const packagePath = path.join(__dirname, '../package.json');
            if (fs.existsSync(packagePath)) {
                const packageJson = JSON.parse(fs.readFileSync(packagePath, 'utf8'));
                console.log('✓ package.json exists');
                console.log(`  - Node.js version: ${process.version}`);
                console.log(`  - Dependencies: ${Object.keys(packageJson.dependencies || {}).length}`);
            } else {
                console.log('✗ package.json not found');
            }
        } catch (error) {
            console.log(`✗ Package dependencies error: ${error.message}`);
        }
    }

    async runAllTests() {
        console.log('Call Center System Test (Node.js + AEP)');
        console.log('=' * 50);
        
        await this.testDatabaseConnection();
        await this.testGoogleSpeech();
        await this.testAEPConfiguration();
        await this.testNodeJSAGI();
        await this.testPackageDependencies();
        
        console.log('\n' + '=' * 50);
        console.log('Test completed!');
        
        // Close database connection
        await this.database.close();
    }
}

// Run tests if this file is executed directly
if (require.main === module) {
    const tester = new SystemTester();
    tester.runAllTests().catch(error => {
        console.error('Test error:', error);
        process.exit(1);
    });
}

module.exports = SystemTester;

File: test_system.py
#!/usr/bin/env python3
"""
Test script for the Call Center system
"""

import sys
import os
sys.path.append(os.path.dirname(os.path.abspath(__file__)))

from config.database import Database
from config.google_speech import GoogleSpeechService
import tempfile

def test_database_connection():
    """ทดสอบการเชื่อมต่อฐานข้อมูล"""
    print("Testing database connection...")
    try:
        db = Database()
        if db.connection:
            print("✓ Database connection successful")
            
            # ทดสอบการค้นหาสาขา
            result = db.get_branch_extension("กรุงเทพ")
            if result:
                print(f"✓ Branch search successful: {result}")
            else:
                print("✗ Branch search failed")
            
            db.close()
        else:
            print("✗ Database connection failed")
    except Exception as e:
        print(f"✗ Database error: {e}")

def test_google_speech():
    """ทดสอบ Google Speech API"""
    print("\nTesting Google Speech API...")
    try:
        gs = GoogleSpeechService()
        
        # ทดสอบ Text-to-Speech
        print("Testing Text-to-Speech...")
        audio_data = gs.text_to_speech("สวัสดีครับ ยินดีต้อนรับสู่ระบบ Call Center")
        if audio_data:
            print("✓ Text-to-Speech successful")
            
            # บันทึกไฟล์เสียงเพื่อทดสอบ
            with tempfile.NamedTemporaryFile(suffix='.wav', delete=False) as f:
                f.write(audio_data)
                print(f"✓ Audio file saved: {f.name}")
        else:
            print("✗ Text-to-Speech failed")
        
        # ทดสอบ Speech-to-Text (ต้องมีไฟล์เสียง)
        print("\nTesting Speech-to-Text...")
        print("Note: Speech-to-Text requires audio file input")
        
    except Exception as e:
        print(f"✗ Google Speech error: {e}")

def test_agi_script():
    """ทดสอบ AGI script"""
    print("\nTesting AGI script...")
    try:
        # ตรวจสอบว่าไฟล์ AGI script มีอยู่
        agi_path = "/var/lib/asterisk/agi-bin/call_routing.py"
        if os.path.exists(agi_path):
            print("✓ AGI script exists")
            
            # ตรวจสอบ permissions
            if os.access(agi_path, os.X_OK):
                print("✓ AGI script is executable")
            else:
                print("✗ AGI script is not executable")
        else:
            print("✗ AGI script not found")
            
    except Exception as e:
        print(f"✗ AGI script error: {e}")

def test_asterisk_config():
    """ทดสอบการตั้งค่า Asterisk"""
    print("\nTesting Asterisk configuration...")
    try:
        config_files = [
            "/etc/asterisk/extensions.conf",
            "/etc/asterisk/sip.conf",
            "/etc/asterisk/manager.conf",
            "/etc/asterisk/modules.conf"
        ]
        
        for config_file in config_files:
            if os.path.exists(config_file):
                print(f"✓ {config_file} exists")
            else:
                print(f"✗ {config_file} missing")
                
    except Exception as e:
        print(f"✗ Asterisk config error: {e}")

def main():
    """ฟังก์ชันหลักสำหรับทดสอบระบบ"""
    print("Call Center System Test")
    print("=" * 50)
    
    # ทดสอบการเชื่อมต่อฐานข้อมูล
    test_database_connection()
    
    # ทดสอบ Google Speech API
    test_google_speech()
    
    # ทดสอบ AGI script
    test_agi_script()
    
    # ทดสอบการตั้งค่า Asterisk
    test_asterisk_config()
    
    print("\n" + "=" * 50)
    print("Test completed!")

if __name__ == "__main__":
    main()